tools/SourceKit/cmake/modules/AddSwiftSourceKit.cmake

```diff
@@ -70,15 +70,6 @@ function(add_sourcekit_default_compiler_flags target)
     # TODO(compnerd) this should really use target_compile_options but the use
     # of keyword and non-keyword flags prevents this
     list(APPEND c_compile_flags "-fblocks")
-    # TODO(compnerd) this should really use target_link_libraries but the use of
-    # explicit_llvm_config using target_link_libraries without keywords on
-    # executables causes conflicts here
-    list(APPEND link_flags "-L${SWIFT_PATH_TO_LIBDISPATCH_BUILD}")
-    list(APPEND link_flags "-lBlocksRuntime")
-    # NOTE(compnerd) since we do not use target_link_libraries, we do not get
-    # the implicit dependency tracking.  Add an explicit dependency until we can
-    # use target_link_libraries.
-    add_dependencies(${target} BlocksRuntime)
```

Comment:
Not sure if this should be removed.  This should be adjusted for the new layout right?  Or am I missing something?

Reply:
BlocksRuntime is statically linked into libdispatch already so it's not correct to link BlocksRuntime again, this build should be linking dispatch to get BlocksRuntime symbols.

```
$ readelf --dyn-syms -W swift-4.2-RELEASE-ubuntu14.04/usr/lib/swift/linux/libdispatch.so | grep -E '(Block_copy|Block_release)'
   605: 0000000000060220   196 FUNC    GLOBAL DEFAULT   11 _Block_release
   740: 00000000000600f0     7 FUNC    GLOBAL DEFAULT   11 _Block_copy
   853: 0000000000060340   201 FUNC    GLOBAL DEFAULT   11 _Block_copy_collectable
```

Reply:
Yes, it is statically linked, but the symbols are not supposed to be re-exported.  This is something that I can look into fixing: see https://github.com/apple/swift/pull/17451 and https://github.com/apple/swift-corelibs-libdispatch/pull/362.  This actually is a bigger issue where the blocks usage breaks down :-(.

Reply:
There should only be one copy of BlocksRuntime in the entire program. For example [dispatch_async](https://developer.apple.com/documentation/dispatch/1453057-dispatch_async) calls `Block_copy` and `Block_release` on the blocks that are passed to it. A program can also call these functions on the same block objects you pass to dispatch_async; if they are different you risk undefined behavior. So BlocksRuntime is really a public part of the libdispatch ABI and statically linking it into dispatch re-exporting its symbols isn't possible.

Reply:
Agreed that there should be a single copy, which is why we need the shared build of it for other targets as well and that was the motivation for the original change that I got distracted from.  I suppose it is high time that I revisit that.

I really think that exporting it as part of libdispatch is fundamentally the wrong approach.  You *can* use block with C, and in such a scenario, you shouldn't pull in libdispatch.  I think it is better to actually fix the build so that we build BlocksRuntime shared and use that appropriately.